text/0000-deprecate-authors-field.md

```diff
@@ -0,0 +1,128 @@
+# RFC: Deprecate the authors field
+
+- Feature Name: `deprecate_authors_field`
+- Start Date: 2021-01-07
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to deprecate the `package.authors` field of `Cargo.toml`.
+This also implies preventing Cargo from auto-filling it, allowing crates to be
+published to crates.io without the field being present, and avoiding displaying
+its contents on the crates.io and docs.rs UI.
+
+# Motivation
+[motivation]: #motivation
+
+The crates.io registry does not allow users to change the contents of already
+published versions: this is highly desirable to ensure working builds don't
+break in the future, but it also has the unfortunate side-effect of preventing
+people from updating the list of crate authors defined in `Cargo.toml`'s
+`package.authors` field.
+
+This is especially problematic when people change their name or want to remove
+their name from the Internet, and the crates.io team doesn't have any way to
+address that at the moment except for deleting the affected crates or versions
+altogether. We don't do that lightly, but there were a few cases where we were
+forced to do so.
+
+The contents of the field also tend to scale poorly as the size of a project
+grows, with projects either making the field useless by just stating "The
+$PROJECT developers" or only naming the original authors without mentioning
+other major contributors.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+crates.io will allow publishing crates without the `package.authors` field, and
+it will stop showing the contents of the field in its UI (the current owners
+will still be shown). docs.rs will also replace that data with the crate
+owners.
+
+`cargo init` will stop pre-populating the field when running the command, and
+it will not include the field at all in the default `Cargo.toml`. Crate authors
+will still be able to manually include the field before publishing if they so
+choose, even though Cargo will warn when trying to publish those crates.
```

Comment:
I'm not fully sold that a warning is necessary. If it's not populated by default, what's the issue with someone adding it if they want to?

Reply:
The main purpose I see for a warning is telling people who ran `cargo init`/`cargo new` before this RFC that they can actually remove it if they so choose. Also, if the goal is to *deprecate* the field we should eventually have people stop using it.

Reply:
I **like** the authors field and would not appreciate it going away.

It's fair to make it not required, but it shouldn't actually be removed entirely.

Reply:
Repeating from the Zulip. I'm not sure I see the need for the deprecation. This field has valid uses in many areas as people have pointed out, if crates.io doesn't want to display it they don't have to, but I don't see why that requires deprecation on Cargo's side, rather than solely making it optional. crates.io could stop displaying the information already even without this RFC.

Reply:
The `[package]` table of `Cargo.toml` only contains metadata used by Cargo or registries, and I'd like to see it remain that way (by deprecating and eventually removing through editions). Ultimately that's not my call to make though, and I'm curious what the Cargo team thinks about it.

Reply:
I am a fan of deprecation. The fields of a manifest file in a package management system are precious, and their numbers tend to trend towards infinite if there's not a clear effort to keep them tidy and minimal (see also package.json). The harm here may not be obvious; many fields can all have many useful reasons to exist. But for new and existing users alike, reading existing manifest files and creating new ones, an excess of optional fields is at best overwhelming and at worst actively confusing. The optionality of a field is not local to the individual manifest file and a name like "author" sounds... authoritative. In my opinion, the cost of keeping this field doesn't outweigh the benefit of a streamlined set of fields. I also think using this as a great, low-risk opportunity to develop a workflow and practice for deprecating manifest fields, which will be critical for the future ergonomics of crates.io, cargo, and devtools that leverage Cargo.toml.

Reply:
> the benefit of keeping this field doesn't outweigh the cost of the longterm maintenance (which will be costly on several axis).

The action with least maintenance is to entirely remove all code involving the `authors` field, and treating it as an unknown field (impossible due to `$CARGO_PKG_AUTHORS` but let's put this aside). 

Currently, both cargo and crates.io *ignores* these unknown fields in `[package]`. So, at least on the axis of code maintenance, actively throwing a deprecation warning is more costly than doing nothing. Unless rust-lang/cargo#3576 is implemented.